core/primitives/address.hpp

```diff
@@ -3,16 +3,146 @@
+#include <string>
+#include <type_traits>
+#include <vector>
+
+#include "boost/variant.hpp"
+#include "common/outcome.hpp"
 
 namespace fc::primitives {
+  using Uint = uint64_t;            // TODO:
+  using Byte = uint8_t;             // consider moving
+  using Bytes = std::vector<Byte>;  // commonly used types
+  using String = std::string;       // to a better place
+};                                  // namespace fc::primitives
+
+namespace fc::primitives::address {
+
+  /**
+   * @class Potential errors creating and handling Filecoin addresses
+   */
+  enum class AddressError : int {
+    UNKNOWN_PROTOCOL = 1, /**< Unknown Address protocol/type */
+    INVALID_PAYLOAD,      /**< Invalid data for a given protocol */
+    UNKNOWN_NETWORK       /**< Unknown network: neither testnet nor mainnet */
+  };
+
+  enum class Network : int { testnet = 0, mainnet };
+
+  enum class Protocol : int { ID = 0, SECP256K1, Actor, BLS };
+
+  template <typename T>
+  constexpr auto to_underlying(T v) noexcept {
+    return static_cast<std::underlying_type_t<T>>(v);
+  }
+
+  struct PublicKeySecp256k1Hash {
+    PublicKeySecp256k1Hash(const Bytes &h) : hash(h) {}
+    PublicKeySecp256k1Hash(const Bytes &&h) : hash(h) {}
+
+    Bytes hash;
+    static const size_t size = 20;
+  };
+
+  struct ActorExecHash {
+    ActorExecHash(const Bytes &h) : hash(h) {}
+    ActorExecHash(const Bytes &&h) : hash(h) {}
+
+    Bytes hash;
+    static const size_t size = 20;
+  };
+
+  struct PublicKeyBLSHash {
+    PublicKeyBLSHash(const Bytes &h) : hash(h) {}
+    PublicKeyBLSHash(const Bytes &&h) : hash(h) {}
+
+    Bytes hash;
+    static const size_t size = 48;
+  };
+
+  bool operator==(const PublicKeySecp256k1Hash &lhs,
+                  const PublicKeySecp256k1Hash &rhs);
+  bool operator==(const ActorExecHash &lhs, const ActorExecHash &rhs);
+  bool operator==(const PublicKeyBLSHash &lhs, const PublicKeyBLSHash &rhs);
 
+  /**
+   * @brief Address interface declaration
+   */
   class Address {
+    using Payload = boost::
+        variant<Uint, PublicKeySecp256k1Hash, ActorExecHash, PublicKeyBLSHash>;
+
+   public:
+    // Creates an empty Address
+    Address() : networkId(Network::testnet), data(Uint{0}), empty(true) {}
+
+   private:
+    Address(Network net, Uint id) : networkId(net), data(id), empty(false) {}
+    Address(Network net, PublicKeySecp256k1Hash &&id)
+        : networkId(net), data(id), empty(false) {}
+    Address(Network net, ActorExecHash &&id)
+        : networkId(net), data(id), empty(false) {}
+    Address(Network net, PublicKeyBLSHash &&id)
+        : networkId(net), data(id), empty(false) {}
+
+    Bytes toBytes();
+
    public:
-    //TODO (kamilsa): FIL-21 Implement primitives
+    // Factory method to create Address object
+    static outcome::result<std::unique_ptr<Address>> makeAddress(
+        Network net, Byte protocol, const Bytes &payload);
+
+    // TODO: Implement creation of an Address by decoding a String (if
```

Comment:
For every TODO please create the task in Jira, so that we do not forget about that

Reply:
This is a consequence of ambiguous spec for the Address component which is introduced in two places in the document itself and has the 3rd "source of truth" in the reference implementation. Until the spec has been completely solidified we won't know what the final interface is going to be. It is possible that this method will just go because it is not in the most latest spec variation and I don't see a use case for it. I'm not sure a Jira task is appropriate

Reply:
I am afraid this will be lost without task